src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs

```diff
@@ -0,0 +1,497 @@
+        if (!ScraperState.FindMagic(bytes, out var state))
+        {
+            return false;
+        }
+        if (Verbose) {
+            Console.WriteLine($"{inputPath}: magic at {state.MagicStart}");
+        }
+        var header = ReadHeader(state);
+        if (Verbose) {
+            DumpHeaderDirectory(header);
+        }
+        var content = ReadContent(state, header);
+        content.AddToModel(_builder);
+        return true;
+    }
+
+    private sealed class ScraperState
+    {
+        public ReadOnlyMemory<byte> Data { get; }
+        public bool LittleEndian { get; }
+        private long _position;
+
+        public long MagicStart => HeaderStart - MagicLE.Length;
+        public long HeaderStart { get; }
+
+        private ScraperState(ReadOnlyMemory<byte> data, bool isLittleEndian, long headerStart)
+        {
+            Data = data;
+            LittleEndian = isLittleEndian;
+            HeaderStart = headerStart;
+            _position = headerStart;
+        }
+
+        public static bool FindMagic(ReadOnlyMemory<byte> bytes, [NotNullWhen(true)] out ScraperState? scraperState)
+        {
+            if (FindMagic(bytes.Span, out int offset, out bool isLittleEndian))
+            {
+                scraperState = new ScraperState(bytes, isLittleEndian, offset + MagicLE.Length);
+                return true;
+            }
+            scraperState = null;
+            return false;
+        }
+
+        private static bool FindMagic(ReadOnlySpan<byte> buffer, out int offset, out bool isLittleEndian)
+        {
+            int start = buffer.IndexOf(MagicLE.Span);
+            if (start != -1)
+            {
+                offset = start;
+                isLittleEndian = true;
+                return true;
+            }
+            start = buffer.IndexOf(MagicBE.Span);
+            if (start != -1)
+            {
+                offset = start;
+                isLittleEndian = false;
+                return true;
+            }
+            offset = 0;
+            isLittleEndian = false;
+            return false;
+        }
+
+        public ulong GetUInt64(long offset) => LittleEndian ? BinaryPrimitives.ReadUInt64LittleEndian(Data.Span.Slice((int)offset)) : BinaryPrimitives.ReadUInt64BigEndian(Data.Span.Slice((int)offset));
+        public uint GetUInt32(long offset) => LittleEndian ? BinaryPrimitives.ReadUInt32LittleEndian(Data.Span.Slice((int)offset)) : BinaryPrimitives.ReadUInt32BigEndian(Data.Span.Slice((int)offset));
+        public ushort GetUInt16(long offset) => LittleEndian ? BinaryPrimitives.ReadUInt16LittleEndian(Data.Span.Slice((int)offset)) : BinaryPrimitives.ReadUInt16BigEndian(Data.Span.Slice((int)offset));
+        public byte GetByte(long offset) => Data.Span[(int)offset];
+
+
+        public void GetBytes(long offset, Span<byte> buffer) => Data.Span.Slice((int)offset, buffer.Length).CopyTo(buffer);
```

Comment:
```suggestion
        public void GetBytes(long offset, ref Span<byte> buffer) => Data.Span.Slice((int)offset, buffer.Length).CopyTo(buffer);
```

Is this correct without the `ref`?

Reply:
> Is this correct without the `ref`?

Yes. `buffer` is a pair of a pointer and a length. `Span<T>.CopyTo()` doesn't modify where `buffer` points or how big it is.